### DISCUSSION TITLE
Jackson 3.0.0 default for `MapperFeature.DEFAULT_VIEW_INCLUSION`

### DISCUSSION THREAD

**Discussion by cowtowncoder** [FasterXML/jackson-future-ideas#75] [Polls]
(part of #72 uber-Discussion)

As of Jackson 2.x, `MapperFeature.DEFAULT_VIEW_INCLUSION` defaults to `true` for backwards-compatibility. This means that by default, properties not annotated with any `@JsonView` are INCLUDED.

A suggestion has been made:

https://github.com/FasterXML/jackson-databind/issues/1484

to change it to `false` instead, so only explicitly annotated properties are included.
We cannot make the change in Jackson 2.x, but could change it for 3.0.0.

So let's see if there is consensus on whether to make proposed change.

## Note on Framework defaults

* Spring: defaults to `false` (i.e. different from Jackson 2.x defaults)

**Comment by sdeleuze** [comment]
I voted to change it to `false`, that makes sense for most use cases IMO and would align with Spring defaults.

**Reply by JooHyukKim** [reply]
Thanks!

**Reply by cowtowncoder** [reply]
Thank you for the explanation @sdeleuze. In fact, I have been wondering about how to collaborate with frameworks more close, since I think defaults of Spring and others are form of "voting" too -- maintainers have determined which defaults they think make most sense, likely based on feedback by their own users too.
So +1 for both voting and explaining rationale.

**Reply by JooHyukKim** [reply]
Got me thinking it may be worth asking other frameworks/library teams about their customized-default configurations and their future plans for change as well. Then count those as votes after informing them.

**Reply by cowtowncoder** [reply]
That'd be possible but for now but maybe it'd be good to just add separate list of bulletin points for framework defaults in description, where known. Will do that wrt Spring here.

**Reply by sdeleuze** [reply]
Yeah and the customizations can vary in a Framework depending on use cases, so I think Jackson should focus on the outstandingly popular one, and this one is a good example IMO. See [this old blog post](https://spring.io/blog/2014/12/02/latest-jackson-integration-improvements-in-spring) where I explain the rationale of this Spring + Jackson feature that requires setting `MapperFeature.DEFAULT_VIEW_INCLUSION` to `false` to make sense and that has become very popular.

**Comment by cowtowncoder** [comment]
Although number of votes is (regrettably) low, results are clear so I will call this: Decision == change the default.

**Comment by cowtowncoder** [comment]
I will re-open this: I have concerns about performance impact, and this is why change has not been merged (there is PR).

Will do some research via `jackson-benchmarks` to hopefully determine how important I think impact is; last time I checked, I think it added ~10% overhead for regular serialization where no View processing was enabled by test (but Jackson internals cannot ignore it due to defaults & the way view processing works).

**Reply by cowtowncoder** [reply]
After some testing I conclude my recollection was wrong: while there may be small amount of overhead for deserialization, it is hardly measurable (see https://github.com/FasterXML/jackson-databind/issues/1484#issuecomment-2841285535 ).
So will close again, proceed with change to defaults.

**Comment by jonenst** [comment]
hi @cowtowncoder, maybe this change is a good opportunity to give guidelines/documentation around views ? It feels to me that people hav different expectations about them, and I think that's why it was/is difficult to chose a default value for DEFAULT_VIEW_INCLUSION ?

**Reply by JooHyukKim** [reply]
<img width="892" height="650" alt="image" src="https://github.com/user-attachments/assets/b660acf1-e872-428c-ad4b-bf3af47e43d3" />

Just looking at the volume on JavaDoc itself, there could probably be more 🤔 
@jonenst Could you point us to any more references? Who, and where?

**Reply by jonenst** [reply]
Hi, here are the resources I found when I was looking around. I also started (1% complete) writing my own guide/summary to try and better understand the current state, because I could not understand how people would want DEFAULT_VIEW_INCLUSION=false. I'll put a short description of my use case at the end to hopefully show how different legitimate use cases may be totally incompatible.

 If you want maybe I can work more on a guide ? But I don't feel particulary legitmate to write it, I don't know much about views except for the few debugging sessions I had with them.

Resources:
inital view feature implementation:
- https://web.archive.org/web/20150510075525/http://jira.codehaus.org/browse/JACKSON-138
Added DEFAULT_VIEW_INCLUSION:
- https://web.archive.org/web/20150510075707/http://jira.codehaus.org/browse/JACKSON-232

docs:
- a single line in https://github.com/FasterXML/jackson-annotations/wiki/Jackson-Annotations
- a paragraph in https://github.com/FasterXML/jackson-docs/wiki/Presentation-Jackson-2.0#jsonview-for-deserialization
- changelogs: https://github.com/FasterXML/jackson/wiki/Jackson-Release-2.9


Things to document according to me...
- view subclassing, advantages and drawbacks
- pojo subclassing and multiple @JsonView at different levels of the POJO class hierarchy 
- pojo composition
- flexibility of configuring objectmapper VS guaranteeing behavior with annotations
- ...

Example: my use case
I have a REST json endpoint returning wrapped data:
{
resourceId: XX
type: YY
resourceData: {...}
}
I use individual POJOs for different resource types, and I use a single wrapper POJO as the root wrapper (this could also be something like the JSON:API wrapper)
individual typed resource data are annotated with type specific views. No views at all on th root wrapper, the views depend on the specific types. When serializing the root wrapper with writerForView(specificTypeViewA.class), with DEFAULT_VIEW_INCLUSION=false, the root wrapper is entirely filtered out, the resourceData field is not recursively inspected, this design feels 100% incompatible with DEFAULT_VIEW_INCLUSION (only solution, annotate the root object with literally all the possible views). Writing annotations that are 100% incompatible with a default objectmapper configuration gives the impression that it's not the right way to do it, so official guides to know the supported use cases for views would help.

**Reply by cowtowncoder** [reply]
Quick note: if your usage relies on `DEFAULT_VIEW_INCLUSION = true`, please do use that setting! There is nothing wrong in using that. It was the original setting after all, based on what I back then thought makes sense too.

As far as I understand, wish for `DEFAULT_VIEW_INCLUSION = false` default comes by users whose use case relies on that (obviously), and the main concern is that some detectable properties that were not meant to be included would accidentally be included. Sort of use cases where explicit `@JsonProperty` is desired (and default getter/setter visibility is set to "none, always require annotation).
So it is more of a security concern (I think), accidental exposure of non-view-associated properties.

On documentation: we would definitely be glad to get more documentation in whatever form makes sense; and can proof-read to align our understandings of how things are expected to work.

**Reply by jonenst** [reply]
> Quick note: if your usage relies on DEFAULT_VIEW_INCLUSION = true, please do use that setting! 

Not sure if there's an established decision on this, but to me, it feels like configuring things in the mapper is for cases where the class author consciously wants to allow a higher level to control/tweak/customize the output (and maybe additionally as an easy escape hatch when you can't change the class annotations and they didn't declare a behavior). And for other cases you annotate the class/fields ? This strategy allows to share a global common objectmapper (like in springboot), make it easy and less errorprone to participate in global config tweaks (like date formats or non null inclusion), but still allow to have strong guarantees on some parts.


The alternative to me is that every class encapsulate its own ObjectMapper (because they want to control the configuration) which is more heavyweight (even if configuration can be cloned). Is it a project goal that most use cases can be supported with a single objectmapper and specialized annotation at the class/field level ?

**Comment by sdeleuze** [comment]
> So it is more of a security concern (I think), accidental exposure of non-view-associated properties.

Security concern indeed, but it is also commonly used to prevent too big workloads being serialized by allowing contextual serialization profiles, so IMO the main reason for the switch is a much better and predictable DevXP for developers to only include what has been explicitly annotated for the use case I described in https://spring.io/blog/2014/12/02/latest-jackson-integration-improvements-in-spring#json-views.

**Comment by jonenst** [comment]
Hi,
for my use case (generic wrapper object + views only inside the wrapped data), I think allowing to annotate the wrapper class to make the behavior independent of the objectmapper config would be useful (you can't do it with JsonView(Something.class) because you'd need a class that extends anything (like the "bottom" type) but it doesn't exist in java, only null can be assigned to anything but not at the type level) (also the bottom type is probably confusing to a lot of users).

I think it's already possible in many case to annotate a class to make a behavior independant of the objectmapper configuration, so this would make it more homogeneous? Also in this case, contrary to cases where the behavior configured at the objectmapper level leads to multiple useful bevahiors (like global date formats or global null inclusion), this one has one behavior that makes the class effectively useless (any view => always empty), so there is an even stronger reason to allow defining this at the class level ?

**Reply by cowtowncoder** [reply]
Annotation applicable to value classes would make sense, conceptually. Would override mapper configuration.
Probably need new annotation type. You could file a new issue (RFE) for jackson-databind for such functionality -- annotation part would go in jackson-annotations but handling in databind.

**Reply by jonenst** [reply]
Thanks ! https://github.com/FasterXML/jackson-databind/issues/5623

**Reply by jonenst** [reply]
Wait, I think I made a mistake. the view selected at runtime (writerWithView) will trigger inclusion of all fields/classes annotated with any superclass ( activeView == null || _view.isAssignableFrom(activeView) ), not the other way around. So you don't need a declaration `@JsonView(Bottom.class)` to include for all views, you need `@JsonView(Object.class)` ! And so it already works, no ?

So this becomes only a question of whether or not the API can be made easier to understand for users ?

**Comment by jonenst** [comment]
Regarding the security aspects of this, it feels like if security is truely important, it should not be a configuration of the object mapper (data can change at runtime), it should be a method of the objectmapper (static enforcement): mapper.writerWithSecureView (need to find an actual good name here) (In springboot, this would probably map to an annotation on the controller)

**Reply by cowtowncoder** [reply]
FWTW this setting unfortunately cannot be used on per-call basis: it changes configuration and construction of serializers which get cached. So as things are, it must be property of mapper and not ObjectWriters.

This is not something that will be changing in near future and likely ever.

**Reply by jonenst** [reply]
Hi, I see thanks for the clear explanation and reality check. So for static enforcement you'd need a new type SecureObjectMapper I guess which would ignore all "unsecure" configuration. But "security" would have to be very important to justify introducing this.

In the absence of static enforcement, maybe the next best thing is a runtime behavior in jackson of "one way" or "fuse" for security related configurations, which would help people (e.g. springboot) implement shared configurable yet secure systems (otherwise everyone must reimplement this kind of behavior on top of a private objectmapper) ? But again, is security an important enough feature to justify this ?

Of course it is not possible to have hard guarantees for a system using only java classes visibility boundaries inside the same jvm process as an enforcement mechanism (bypassable with reflection, dynamic code generation etc.), so this is all to protect only against mistakes or "innocent looking" abuse, which again raises the question of the importance of security in this design.

All in all I think this just means that the security argument for the new default is not a very strong one. In fact I believe the new default is slightly worse for security. I think more accidents will happen with the new default.

Compare the following two scenarios:
Previous default:
- Developer A annotates class with JsonView, expecting unanotated fields to always be excluded. They test, they see that they are *not* excluded. They set DEFAULT_VIEW_INCLUSION explictely to false ( Note: here they could have annotated their class with JsonView(Hidden.class), ensuring the fields would never be included by accident. this should be in the guide)
- developer B comes later and adds another class which requires DEFAULT_VIEW_INCLUSION true, they will see that it has been explicitely set to false by A and will presumably learn about the fields that A wanted to exclude.

   
New default:
- Developer A annotates class with JsonView, expecting unanotated fields to always be excluded. They test, it works by default, they move on
- developer B comes later and adds another class which requires DEFAULT_VIEW_INCLUSION true, they see that the project is not explicitely setting DEFAULT_VIEW_INCLUSION (so using the default value), so they may think "this was unset, so probably nothing requires it, I'll just set it to true". And now the fields of the other class are exposed.

I've written a lot about security which is not even my main use case and not very different between the two scenarios, I guess I'll leave it at that to avoid wasting more of your time unless you think this discussion should continue ?

Note: The security argument does not take away the fact that the new default is better for people who implicitely do inclusions/exclusions based on exhaustive whitelists, which is to me the strong argument in favor of the new default (and implicit behavior doesn't play well with security)

**Comment by jonenst** [comment]
Hi,
currently, with DEFAULT_VIEW_INCLUSION=false, when you annotate à String field with @JsonView(XXX) the data gets included in the view. But if you annoate a Pojo field with @JsonView(XXX), you get an empty object in the view. You have to additionally go and annotate the POJO fields (or class) with the same View to have any data. And again recursively for any field that is a POJO. For big hierarchies this becomes very error prone, and missing fields/classes in the whitelist will "cut off" whole subtrees.

To alleviate this difficulty a possibility that seems achievable with the restriction of a an implementation that can map this behavior to constructed cached serializers, would be to have classes with 0 JsonView annotations be considered unfiltered, all fields always included (like in DEFAULT_VIEW_INCLUSION=true)

It could either be the new behaviorr of DEFAULT_VIEW_INCLUSION, or if backwards compatibility or the previous behavior is important, maybe a new mapper configuration DEFAULT_VIEW_SCOPE with values ANNOTATED_ONLY and ALL saying whether the DEFAULT_VIEW_INCLUSION config applies to fields in classes with at last one defined JsonView annotation or to all fields in any class ?

A downside of this approach is the increase in complexity: adding the first annotation on a field in a class changes the behavior for other fields in this class. But maybe it's worth it to make DEFAULT_VIEW_INCLUSION more usefull with big hiearchies ?

**Reply by JooHyukKim** [reply]
@jonenst Could you file a new feature request issue in databind module https://github.com/FasterXML/jackson-databind ?

This is closed discussion about default value of `MapperFeature.DEFAULT_VIEW_INCLUSION`, discussing whether it be `true`/`false`, NOT about HOW `@JsonView` behave.
Thanks!
